cpp/src/replace/replace.cu

```diff
@@ -13,31 +16,528 @@
+   }
+
+  /* --------------------------------------------------------------------------*/
+  /**
+   * @brief Kernel that replaces elements from `output_data` given the following
+   *        rule: replace all `values_to_replace[i]` in [values_to_replace_begin`,
+   *        `values_to_replace_end`) present in `output_data` with `d_replacement_values[i]`.
+   *
+   * @tparam input_has_nulls `true` if output column has valid mask, `false` otherwise
+   * @tparam replacement_has_nulls `true` if replacement_values column has valid mask, `false` otherwise
+   * The input_has_nulls and replacement_has_nulls template parameters allows us to specialize
+   * this kernel for the different scenario for performance without writing different kernel.
+   *
+   * @param[in] input_data Device array with the data to be modified
+   * @param[in] input_valid Valid mask associated with input_data
+   * @param[out] output_data Device array to store the data from input_data
+   * @param[out] output_valid Valid mask associated with output_data
+   * @param[out] output_valid_count #valid in output column
+   * @param[in] nrows # rows in `output_data`
+   * @param[in] values_to_replace_begin Device pointer to the beginning of the sequence
+   * of old values to be replaced
+   * @param[in] values_to_replace_end  Device pointer to the end of the sequence
+   * of old values to be replaced
+   * @param[in] d_replacement_values Device array with the new values
+   * @param[in] replacement_valid Valid mask associated with d_replacement_values
+   *
+   * @returns
+   */
+  /* ----------------------------------------------------------------------------*/
+  template <class T,
+            bool input_has_nulls, bool replacement_has_nulls>
+  __global__
+  void replace_kernel(cudf::column_device_view input,
+                      cudf::mutable_column_device_view output,
+                      cudf::size_type * __restrict__    output_valid_count,
+                      cudf::size_type                   nrows,
+                      cudf::column_device_view values_to_replace,
+                      cudf::column_device_view replacement)
+  {
+  const T* __restrict__ input_data = input.data<T>();
+  cudf::bitmask_type const * __restrict__ input_valid = input.null_mask();
+  T * __restrict__ output_data = output.data<T>();
+  cudf::bitmask_type * __restrict__ output_valid = output.null_mask();
+  const T* __restrict__ values_to_replace_begin = values_to_replace.data<T>();
+  const T* __restrict__ values_to_replace_end = values_to_replace.data<T>() + values_to_replace.size();
+  const T* __restrict__ d_replacement_values = replacement.data<T>();
+  cudf::bitmask_type const * __restrict__ replacement_valid = replacement.null_mask();
+
+  cudf::size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+
+  uint32_t active_mask = 0xffffffff;
+  active_mask = __ballot_sync(active_mask, i < nrows);
+  __shared__ uint32_t valid_sum[cudf::experimental::detail::warp_size];
```

Comment:
This isn't necessary if each warp counts the valid count in a register and then calls `single_lane_block_sum_reduce`.

Reply:
Done.